### PR TITLE
KAFKA-16260: Deprecate window.size.ms in StreamsConfig.java and index.html

### PR DESCRIPTION
**Pull request by Cerchie**
This is the first step towards [KIP-1020](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=290982804)

Pieces of config to deprecate: 
- window.size.ms
- windowed.inner.class.serde

In the code, `WINDOWED_INNER_CLASS_SERDE` was directly below `DEFAULT_WINDOWED_VALUE_SERDE_INNER_CLASS` with no decorators or javadocs. I added a @Deprecated decorator. 

I added a @Deprecated decorator to `WINDOW_SIZE_MS_CONFIG` and a short note regarding deprecation to the javadoc. Normally I'd note the release it was dep'd in as well as alternate config but those are not established yet. 

In the documentation, I added a short note on deprecation to `window.size.ms` but again, had no further info to add.

Something titled `windowed.inner.class.serde` was not present in config-streams.html

### CODE REVIEW

File: .github/workflows/codesee-arch-diagram.yml [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# This workflow was added by CodeSee. Learn more at https://codesee.io/
```
Comment (mjsax) [inline .github/workflows/codesee-arch-diagram.yml]: This file should be remove from the PR.

File: docs/streams/developer-guide/config-streams.html [hunk 2]
```diff
@@ -461,9 +461,9 @@ <h4><a class="toc-backref" href="#id23">num.standby.replicas</a><a class="header
             <td>86400000 milliseconds (1 day)</td>
           </tr>
           <tr class="row-even"><td>window.size.ms</td>
```
Comment (mjsax) [inline docs/streams/developer-guide/config-streams.html]: I believe in your other PR, you added `(Deprecated)` here?

Let's be consistent.

Comment (mjsax) [reply docs/streams/developer-guide/config-streams.html]: About the other config. If it's missing, we should actually add a new row for it

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 3]
```diff
@@ -595,7 +595,8 @@ public class StreamsConfig extends AbstractConfig {
     public static final String DEFAULT_WINDOWED_VALUE_SERDE_INNER_CLASS = "default.windowed.value.serde.inner";
     private static final String DEFAULT_WINDOWED_VALUE_SERDE_INNER_CLASS_DOC = "Default serializer / deserializer for the inner class of a windowed value. Must implement the " +
         "<code>org.apache.kafka.common.serialization.Serde</code> interface.";
-
+    
+    @Deprecated
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:602]: We should add JavaDocs similar to `window.size.ms`.

**Review by mjsax — COMMENTED** [review]
Thanks for the PR @Cerchie 

I think the overall change is small enough to do everything is a single PR. For the release number, I would got with 3.9 for now, as it's the next target release. We can still update the PR and change if necessary.

File: .github/workflows/codesee-arch-diagram.yml (hunk 1 above)
Comment (Cerchie) [reply .github/workflows/codesee-arch-diagram.yml]: done

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java (hunk 3 above)
Comment (Cerchie) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:602]: yes. done.

File: docs/streams/developer-guide/config-streams.html [hunk 4]
```diff
@@ -1022,7 +1019,7 @@ <h4><a class="toc-backref" href="#id31">topology.optimization</a><a class="heade
       <blockquote>
         <div>
           <p>
-            A configuration telling Kafka Streams if it should optimize the topology and what optimizations to apply. Acceptable values are: <code>StreamsConfig.NO_OPTIMIZATION</code> (<code>none</code>), <code>StreamsConfig.OPTIMIZE</code> (<code>all</code>) or a comma separated list of specific optimizations: <code>StreamsConfig.REUSE_KTABLE_SOURCE_TOPICS</code> (<code>reuse.ktable.source.topics</code>), <code>StreamsConfig.MERGE_REPARTITION_TOPICS</code> (<code>merge.repartition.topics</code>), <code>StreamsConfig.SINGLE_STORE_SELF_JOIN</code> (<code>single.store.self.join</code>).
+            A configuration telling Kafka Streams if it should optimize the topology and what optimizations to apply. Acceptable values are: <code>StreamsConfig.NO_OPTIMIZATION</code> (<code>none</code>), <code>StreamsConfig.OPTIMIZE</code> (<code>all</code>) or a comma separated list of specific optimizations: (<code>StreamsConfig.REUSE_KTABLE_SOURCE_TOPICS</code> (<code>reuse.ktable.source.topics</code>), <code>StreamsConfig.MERGE_REPARTITION_TOPICS</code> (<code>merge.repartition.topics</code>)).
```
Comment (Cerchie) [inline docs/streams/developer-guide/config-streams.html:1048]: this came in from trunk

File: docs/streams/developer-guide/config-streams.html [hunk 5]
```diff
@@ -1140,7 +1125,7 @@ <h4><a class="toc-backref" href="#id18">Default Values</a><a class="headerlink"
           <td>Consumer</td>
           <td><code class="docutils literal"><span class="pre">1000</span></code></td>
         </tr>
-        <tr class="row-odd">
+        <tr class="row-even">
```
Comment (Cerchie) [inline docs/streams/developer-guide/config-streams.html]: again, I think these row changes are from trunk

File: docs/streams/developer-guide/config-streams.html [hunk 6]
```diff
@@ -242,29 +240,24 @@ <h4><a class="toc-backref" href="#id23">num.standby.replicas</a><a class="header
...
               of the application.</td>
             <td>the empty string</td>
           </tr>
-          <tr class="row-even"><td>buffered.records.per.partition</td>
+          <tr class="row-odd"><td>buffered.records.per.partition</td>
             <td>Low</td>
             <td colspan="2">The maximum number of records to buffer per partition.</td>
             <td><code class="docutils literal"><span class="pre">1000</span></code></td>
           </tr>
-          <tr class="row-odd"><td>statestore.cache.max.bytes</td>
```
Comment (mjsax) [inline docs/streams/developer-guide/config-streams.html]: Why does we remove this config?

File: docs/streams/developer-guide/config-streams.html [hunk 7]
```diff
@@ -307,12 +300,12 @@ <h4><a class="toc-backref" href="#id23">num.standby.replicas</a><a class="header
               set by the user or all serdes must be passed in explicitly (see also default.key.serde).</td>
             <td><code class="docutils literal"><span class="pre">null</span></code></td>
           </tr>
-          <tr class="row-even"><td>default.windowed.key.serde.inner (Deprecated. Use windowed.inner.class.serde instead.)</td>
+          <tr class="row-even"><td>default.windowed.key.serde.inner</td>
```
Comment (mjsax) [inline docs/streams/developer-guide/config-streams.html]: Why do we remove the deprecation statement?

File: docs/streams/developer-guide/config-streams.html [hunk 8]
```diff
@@ -307,12 +300,12 @@ <h4><a class="toc-backref" href="#id23">num.standby.replicas</a><a class="header
...
             <td>Medium</td>
             <td colspan="2">Default serializer/deserializer for the inner class of windowed keys, implementing the <code class="docutils literal"><span class="pre">Serde</span></code> interface.</td>
             <td><code class="docutils literal"><span class="pre">null</span></code></td>
           </tr>
-          <tr class="row-odd"><td>default.windowed.value.serde.inner (Deprecated. Use windowed.inner.class.serde instead.)</td>
+          <tr class="row-odd"><td>default.windowed.value.serde.inner</td>
```
Comment (mjsax) [inline docs/streams/developer-guide/config-streams.html]: as above

File: docs/streams/developer-guide/config-streams.html [hunk 9]
```diff
@@ -439,28 +431,27 @@ <h4><a class="toc-backref" href="#id23">num.standby.replicas</a><a class="header
...
             <td>300000 milliseconds (5 minutes)</td>
           </tr>
-          <tr class="row-odd"><td>topology.optimization</td>
+          <tr class="row-even"><td>topology.optimization</td>
             <td>Medium</td>
-            <td colspan="2">A configuration telling Kafka Streams if it should optimize the topology and what optimizations to apply. Acceptable values are: <code>StreamsConfig.NO_OPTIMIZATION</code> (<code>none</code>), <code>StreamsConfig.OPTIMIZE</code> (<code>all</code>) or a comma separated list of specific optimizations: <code>StreamsConfig.REUSE_KTABLE_SOURCE_TOPICS</code> (<code>reuse.ktable.source.topics</code>), <code>StreamsConfig.MERGE_REPARTITION_TOPICS</code> (<code>merge.repartition.topics</code>),
-            <code>StreamsConfig.SINGLE_STORE_SELF_JOIN</code> (<code>single.store.self.join</code>). </td>
```
Comment (mjsax) [inline docs/streams/developer-guide/config-streams.html]: Why are we removing this? It's a valid option.

File: docs/streams/developer-guide/config-streams.html [hunk 10]
```diff
@@ -684,7 +675,7 @@ <h4><a class="toc-backref" href="#id9">default.value.serde</a><a class="headerli
             </div></blockquote>
         </div>
         <div class="section" id="default-windowed-key-serde-inner">
-          <h4><a class="toc-backref" href="#id32">default.windowed.key.serde.inner</a><a class="headerlink" href="#default-windowed-key-serde-inner" title="Permalink to this headline"></a> (Deprecated.)</h4>
+          <h4><a class="toc-backref" href="#id32">default.windowed.key.serde.inner</a><a class="headerlink" href="#default-windowed-key-serde-inner" title="Permalink to this headline"></a></h4>
```
Comment (mjsax) [inline docs/streams/developer-guide/config-streams.html]: Why remove the deprecated hint.

File: docs/streams/developer-guide/config-streams.html [hunk 11]
```diff
@@ -697,7 +688,7 @@ <h4><a class="toc-backref" href="#id32">default.windowed.key.serde.inner</a><a c
             </div></blockquote>
         </div>
         <div class="section" id="default-windowed-value-serde-inner">
-          <h4><a class="toc-backref" href="#id33">default.windowed.value.serde.inner</a><a class="headerlink" href="#default-windowed-value-serde-inner" title="Permalink to this headline"></a>(Deprecated.)</h4>
+          <h4><a class="toc-backref" href="#id33">default.windowed.value.serde.inner</a><a class="headerlink" href="#default-windowed-value-serde-inner" title="Permalink to this headline"></a></h4>
```
Comment (mjsax) [inline docs/streams/developer-guide/config-streams.html]: as above

File: docs/streams/developer-guide/config-streams.html [hunk 12]
```diff
@@ -708,6 +699,12 @@ <h4><a class="toc-backref" href="#id33">default.windowed.value.serde.inner</a><a
               <p>This is discussed in more detail in <a class="reference internal" href="datatypes.html#streams-developer-guide-serdes"><span class="std std-ref">Data types and serialization</span></a>.</p>
             </div></blockquote>
         </div>
+      <div class="section" id="windowed-inner-class-serde">
```
Comment (mjsax) [inline docs/streams/developer-guide/config-streams.html:716]: We should add a deprecation hint, as KIP-1020 does deprecate this config, right?

File: docs/streams/developer-guide/config-streams.html (hunk 4 above)
Comment (mjsax) [reply docs/streams/developer-guide/config-streams.html:1048]: Yes, it was another doc fix PR to add a missing option -- we should keep it as is.

File: docs/streams/developer-guide/config-streams.html (hunk 5 above)
Comment (mjsax) [reply docs/streams/developer-guide/config-streams.html]: Well,l the row above this one is "even" so this one should be "odd" -- no need to set it to "even"

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 13]
```diff
@@ -595,7 +595,10 @@ public class StreamsConfig extends AbstractConfig {
     public static final String DEFAULT_WINDOWED_VALUE_SERDE_INNER_CLASS = "default.windowed.value.serde.inner";
     private static final String DEFAULT_WINDOWED_VALUE_SERDE_INNER_CLASS_DOC = "Default serializer / deserializer for the inner class of a windowed value. Must implement the " +
         "<code>org.apache.kafka.common.serialization.Serde</code> interface.";
-
+    
+    /** {@code windowed.inner.class.serde
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]:
```suggestion
    /** {@code windowed.inner.class.serde}
```

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 14]
```diff
@@ -595,7 +595,10 @@ public class StreamsConfig extends AbstractConfig {
...
+     * @deprecated since 3.9} */
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]:
```suggestion
     * @deprecated since 3.9 */
```

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 15]
```diff
@@ -776,7 +779,9 @@ public class StreamsConfig extends AbstractConfig {
         "For any timeout larger than 0ms, a task will retry at least once before an error is raised.";
 
 
-    /** {@code window.size.ms} */
+    /** {@code window.size.ms
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]:
```suggestion
    /** {@code window.size.ms}
```

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 16]
```diff
@@ -776,7 +779,9 @@ public class StreamsConfig extends AbstractConfig {
...
+     * @deprecated since 3.9} */
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]:
```suggestion
     * @deprecated since 3.9 */
```

**Review by mjsax — COMMENTED** [review]
Thanks for updating the PR. It should also add the configs to the new places, ie, `TimeWindowedSerializer`, and `TimeWindowedDesrializer` as agree on the KIP (we should not merge just the deprecation part w/o the replacements to not leave the code base in a partial state)

File: docs/streams/developer-guide/config-streams.html (hunk 6 above)
Comment (Cerchie) [reply docs/streams/developer-guide/config-streams.html]: coming in from trunk. it's not currently on https://kafka.apache.org/documentation/streams/developer-guide/config-streams.html

File: docs/streams/developer-guide/config-streams.html (hunk 7 above)
Comment (Cerchie) [reply docs/streams/developer-guide/config-streams.html]: again, change coming in from trunk

File: docs/streams/developer-guide/config-streams.html (hunk 8 above)
Comment (Cerchie) [reply docs/streams/developer-guide/config-streams.html]: same-- a change from trunk. I would have these would be in there, not sure why they're gone now.

File: docs/streams/developer-guide/config-streams.html (hunk 9 above)
Comment (Cerchie) [reply docs/streams/developer-guide/config-streams.html]: from trunk

File: docs/streams/developer-guide/config-streams.html (hunk 10 above)
Comment (Cerchie) [reply docs/streams/developer-guide/config-streams.html]: from trunk

File: docs/streams/developer-guide/config-streams.html (hunk 2 above)
Comment (Cerchie) [reply docs/streams/developer-guide/config-streams.html]: will need to do a rebase/merge to get it in. wanted to make sure things look good so far before adding the dep'd notice

File: docs/streams/developer-guide/config-streams.html [hunk 17]
```diff
@@ -91,14 +91,12 @@
               <li><a class="reference internal" href="#rocksdb-config-setter" id="id20">rocksdb.config.setter</a></li>
               <li><a class="reference internal" href="#state-dir" id="id14">state.dir</a></li>
               <li><a class="reference internal" href="#topology-optimization" id="id31">topology.optimization</a></li>
-              <li><a class="reference internal" href="#windowed-inner-class-serde" id="id38">windowed.inner.class.serde</a></li>
```
Comment (mjsax) [inline docs/streams/developer-guide/config-streams.html]: The KIP deprecates this config. Is it pre-mature to remove it? Would it be better to just highlight it's deprecated now as we do for other configs?

File: docs/streams/developer-guide/config-streams.html [hunk 18]
```diff
@@ -91,14 +91,12 @@
...
             </ul>
           </li>
           <li><a class="reference internal" href="#kafka-consumers-and-producer-configuration-parameters" id="id16">Kafka consumers and producer configuration parameters</a>
             <ul>
               <li><a class="reference internal" href="#naming" id="id17">Naming</a></li>
               <li><a class="reference internal" href="#default-values" id="id18">Default Values</a></li>
-              <li><a class="reference internal" href="#parameters-controlled-by-kafka-streams" id="id26">Parameters controlled by Kafka Streams</a></li>
```
Comment (mjsax) [inline docs/streams/developer-guide/config-streams.html]: Why remove this line?

File: docs/streams/developer-guide/config-streams.html [hunk 19]
```diff
@@ -265,6 +263,7 @@ <h4><a class="toc-backref" href="#id23">num.standby.replicas</a><a class="header
             <td>10485760</td>
           </tr>
           <tr class="row-even"><td>cache.max.bytes.buffering (Deprecated. Use statestore.cache.max.bytes instead.)</td>
+          <tr class="row-even"><td>cache.max.bytes.buffering</td>
```
Comment (mjsax) [inline docs/streams/developer-guide/config-streams.html]: Why do we insert this line here? Seems like a duplicate of the exiting line right above it?

File: docs/streams/developer-guide/config-streams.html [hunk 20]
```diff
@@ -439,28 +437,30 @@ <h4><a class="toc-backref" href="#id23">num.standby.replicas</a><a class="header
...
-            <code>StreamsConfig.SINGLE_STORE_SELF_JOIN</code> (<code>single.store.self.join</code>). </td>
+              <code>StreamsConfig.SINGLE_STORE_SELF_JOIN</code> (<code>single.store.self.join</code>). </td>
             <td><code> NO_OPTIMIZATION</code></td>
+            <td colspan="2">A configuration telling Kafka Streams if it should optimize the topology and what optimizations to apply. Acceptable values are: <code>StreamsConfig.NO_OPTIMIZATION</code> (<code>none</code>), <code>StreamsConfig.OPTIMIZE</code> (<code>all</code>) or a comma separated list of specific optimizations: (<code>StreamsConfig.REUSE_KTABLE_SOURCE_TOPICS</code> (<code>reuse.ktable.source.topics</code>), <code>StreamsConfig.MERGE_REPARTITION_TOPICS</code> (<code>merge.repartition.topics</code>)). </td>
```
Comment (mjsax) [inline docs/streams/developer-guide/config-streams.html]: Seem to be a duplicate of the line above (incorrect rebase conflict resolution?) (same for next line below)

**Review by mjsax — COMMENTED** [review]
PR shows merge conflicts again... Seems you keep chasing `trunk`... :(

File: docs/streams/developer-guide/config-streams.html [hunk 21]
```diff
@@ -243,19 +243,19 @@ <h4><a class="toc-backref" href="#id23">num.standby.replicas</a><a class="header
           </tr>
           </thead>
           <tbody valign="top">
-          <tr class="row-even"><td>acceptable.recovery.lag</td>
+          <tr class="row-odd"><td>acceptable.recovery.lag</td>
```
Comment (mjsax) [inline docs/streams/developer-guide/config-streams.html:246]: Why do we change this? Seem the top row with "Parameter Name" is `odd` already, so `even` seems to be correct here?

File: docs/streams/developer-guide/config-streams.html [hunk 22]
```diff
@@ -689,7 +689,7 @@ <h4><a class="toc-backref" href="#id9">default.value.serde</a><a class="headerli
             </div></blockquote>
         </div>
         <div class="section" id="default-windowed-key-serde-inner">
-          <h4><a class="toc-backref" href="#id32">default.windowed.key.serde.inner</a><a class="headerlink" href="#default-windowed-key-serde-inner" title="Permalink to this headline"></a> (Deprecated.)</h4>
+          <h4><a class="toc-backref" href="#id32">default.windowed.key.serde.inner</a><a class="headerlink" href="#default-windowed-key-serde-inner" title="Permalink to this headline"></a>(Deprecated.)</h4>
```
Comment (mjsax) [inline docs/streams/developer-guide/config-streams.html:692]: Why remove the space?

File: docs/streams/developer-guide/config-streams.html [hunk 23]
```diff
@@ -713,6 +713,12 @@ <h4><a class="toc-backref" href="#id33">default.windowed.value.serde.inner</a><a
               <p>This is discussed in more detail in <a class="reference internal" href="datatypes.html#streams-developer-guide-serdes"><span class="std std-ref">Data types and serialization</span></a>.</p>
             </div></blockquote>
         </div>
+      <div class="section" id="windowed-inner-class-serde">
+        <h4><a class="toc-backref" href="#id39">windowed.inner.class.serde</a><a class="headerlink" href="#windowed-inner-class-serde" title="Permalink to this headline"></a>(Deprecated.)</h4>
```
Comment (mjsax) [inline docs/streams/developer-guide/config-streams.html:717]:
```suggestion
        <h4><a class="toc-backref" href="#id39">windowed.inner.class.serde</a><a class="headerlink" href="#windowed-inner-class-serde" title="Permalink to this headline"></a> (Deprecated.)</h4>
```

File: docs/streams/developer-guide/config-streams.html [hunk 24]
```diff
@@ -713,6 +713,12 @@ <h4><a class="toc-backref" href="#id33">default.windowed.value.serde.inner</a><a
...
+        <blockquote>
+          <div><p>The default Serializer/Deserializer class for the inner class of a windowed record. Must implement the <code class="docutils literal"><span class="pre">org.apache.kafka.common.serialization.Serde </span></code> interface. Note that setting this config in KafkaStreams application would result in an error as it is meant to be used only from Plain consumer client.</p>
```
Comment (mjsax) [inline docs/streams/developer-guide/config-streams.html:719]:
```suggestion
          <div><p>The default Serializer/Deserializer class for the inner class of a windowed record. Must implement the <code class="docutils literal"><span class="pre">org.apache.kafka.common.serialization.Serde</span></code> interface. Note that setting this config in KafkaStreams application would result in an error as it is meant to be used only from Plain consumer client.</p>
```

File: docs/streams/developer-guide/config-streams.html [hunk 25]
```diff
@@ -1039,7 +1045,7 @@ <h4><a class="toc-backref" href="#id31">topology.optimization</a><a class="heade
       <blockquote>
         <div>
           <p>
-            A configuration telling Kafka Streams if it should optimize the topology and what optimizations to apply. Acceptable values are: <code>StreamsConfig.NO_OPTIMIZATION</code> (<code>none</code>), <code>StreamsConfig.OPTIMIZE</code> (<code>all</code>) or a comma separated list of specific optimizations: <code>StreamsConfig.REUSE_KTABLE_SOURCE_TOPICS</code> (<code>reuse.ktable.source.topics</code>), <code>StreamsConfig.MERGE_REPARTITION_TOPICS</code> (<code>merge.repartition.topics</code>), <code>StreamsConfig.SINGLE_STORE_SELF_JOIN</code> (<code>single.store.self.join</code>).
+            A configuration telling Kafka Streams if it should optimize the topology and what optimizations to apply. Acceptable values are: <code>StreamsConfig.NO_OPTIMIZATION</code> (<code>none</code>), <code>StreamsConfig.OPTIMIZE</code> (<code>all</code>) or a comma separated list of specific optimizations: (<code>StreamsConfig.REUSE_KTABLE_SOURCE_TOPICS</code> (<code>reuse.ktable.source.topics</code>), <code>StreamsConfig.MERGE_REPARTITION_TOPICS</code> (<code>merge.repartition.topics</code>)).
```
Comment (mjsax) [inline docs/streams/developer-guide/config-streams.html:1048]: Why do we remove this option? Seems not related to this PR (and also not necessary -- this option is valid).

**Review by mjsax — COMMENTED** [review]
A few nit.

As I said earlier, I would like to extend this PR to include all changes we need for the KIP.

If you really want to split into multiple PRs, we should first do a PR that is adding the new stuff, before we merge this PR to avoid deprecating things w/o a replacement.

**Comment by mjsax** [comment]
Closing this for now. We can reopen anytime.
